File: HelloWorldApp.scala
package hello

import java.util._

import java.util.stream.Collectors

import Student._

import scala.beans.{BeanProperty, BooleanBeanProperty}

//remove if not needed
import scala.collection.JavaConversions._

object Student {

  @BeanProperty
  var studentCount: Int = 0

}

abstract class Student(@BeanProperty protected var name: String,
                       @BeanProperty var dateOfBirth: Date) {

  { studentCount += 1; studentCount - 1 }

  def this(dateOfBirth: Date) = this(null, dateOfBirth)

  def doWork(): Unit

}

trait GuitarPlayer {

  def playGuitar(): Unit

}

trait PianoPlayer {

  def playPiano(): Unit

}

class MusicStudent(name: String, dateOfBirth: Date)
    extends Student(dateOfBirth)
    with GuitarPlayer
    with PianoPlayer {

  super.setName(name)

  def playGuitar(): Unit = {
    println("Playing Guitar")
  }

  def playPiano(): Unit = {
    println("Playing Piano")
  }

  def doWork(): Unit = {
    this.playGuitar()
    this.playPiano()
  }

}

object HelloWorldApp {

  def main(args: Array[String]): Unit = {
    val student: Student = new MusicStudent("wimal", new Date())
    student.doWork()
// Lists
    val students: List[Student] = new ArrayList[Student]()
    students.add(student)
    val firstStudent: Student = students.get(0)
    firstStudent.doWork()
// Iterators
    val it: Iterator[Student] = students.iterator()
    while (it.hasNext) {
      val next: Student = it.next()
      next.doWork()
    }
// Maps
    val studentMap: Map[String, List[Student]] =
      new HashMap[String, List[Student]]()
    studentMap.put("musicStudents", students)
    val firstMusicStudent: Student = studentMap.get("musicStudents").get(0)
    firstMusicStudent.doWork()
// Arrays
    val studentsArray: Array[Student] = Array.ofDim[Student](1)
    studentsArray(0) = student
    studentsArray(0).doWork()
// 2D Arrays
    val students2DArray: Array[Array[Student]] = Array.ofDim[Student](1, 1)
    students2DArray(0)(0) = student
    println(students2DArray(0)(0).getName)
// Lambda
    val result: List[Student] = students
      .stream()
      .filter((lmbStudent) => ("wimal" == lmbStudent.getName))
      .collect(Collectors.toList())
    println(result.size)
    println(Student.getStudentCount)
  }

}

File: HelloWorldApp.java
package hello;

import java.util.*;
import java.util.stream.Collectors;

abstract class Student {
    private String name;
    private Date dateOfBirth;

    private static int studentCount = 0;

    public Student(Date dateOfBirth) {
        this(null, dateOfBirth);
    }

    public Student(String name, Date dateOfBirth) {
        this.name = name;
        this.dateOfBirth = dateOfBirth;
        studentCount++;
    }

    protected void setName(String name) {
        this.name = name;
    }

    protected String getName() {
        return this.name;
    }

    public Date getDateOfBirth() {
        return this.dateOfBirth;
    }

    public abstract void doWork();

    public static int getStudentCount() {
        return studentCount;
    }
}

interface GuitarPlayer {
    void playGuitar();
}

interface PianoPlayer {
    void playPiano();
}

class MusicStudent extends Student implements GuitarPlayer, PianoPlayer {

    public MusicStudent(String name, Date dateOfBirth) {
        super(dateOfBirth);
        super.setName(name);
    }

    public void playGuitar() {
        System.out.println("Playing Guitar");
    }

    public void playPiano() {
        System.out.println("Playing Piano");
    }

    public void doWork() {
        this.playGuitar();
        this.playPiano();
    }

    @Override
    public String getName() {
        return "Maestro" + super.getName();
    }
}

class HelloWorldApp {
    public static void main(String[] args) throws IllegalArgumentException, NullPointerException {
        try {
            Student student = new MusicStudent("wimal", new Date());
            student.doWork();

            // Lists
            List<Student> students = new ArrayList();
            students.add(student);

            Student firstStudent = students.get(0);
            if (firstStudent == null) {
                throw new IllegalArgumentException("Should have at least one student ...");
            }

            firstStudent.doWork();

            // Iterators
            Iterator<Student> it = students.iterator();
            while (it.hasNext()) {
                Student next = it.next();
                next.doWork();
            }

            // Maps
            Map<String, List<Student>> studentMap = new HashMap();
            studentMap.put("musicStudents", students);

            Student firstMusicStudent = studentMap.get("musicStudents").get(0);
            firstMusicStudent.doWork();

            // Arrays
            Student[] studentsArray = new Student[1];
            studentsArray[0] = student;

            studentsArray[0].doWork();

            // 2D Arrays
            Student[][] students2DArray = new Student[1][1];
            students2DArray[0][0] = student;

            System.out.println(students2DArray[0][0].getName());

            // Lambda
            List<Student> result = students.stream()
                    .filter(lmbStudent -> ("wimal".equals(lmbStudent.getName())))
                    .collect(Collectors.toList());

            System.out.println(result.size());

            System.out.println(Student.getStudentCount());
        } catch (IllegalArgumentException illex) {
            illex.printStackTrace();
            throw illex;
        } catch (NullPointerException nex) {
            nex.printStackTrace();
            throw nex;
        } finally {
            System.out.println("In the finally block");
        }
    }
}

